File: cpp_exercise/backtracking_2_2.py
from typing import List

class Solution:
    """
    用于解决N皇后问题的类。
    N皇后问题要求在N×N的棋盘上放置N个皇后，使得它们彼此不能相互攻击。
    """

    def __init__(self):
        """
        初始化Solution类的实例。
        创建用于记录结果和跟踪列、对角线和反对角线占用情况的数据结构。
        """
        self.result = []
        self.cols = set()
        self.diagonals = set()
        self.back_diagonals = set()
        
    def solveNQueens(self, n: int) -> List[List[str]]:
        """
        解决N皇后问题。
        
        Args:
            n: 棋盘的大小和皇后的数量。
            
        Returns:
            List[List[str]]: 所有可能的N皇后解决方案，每个解决方案表示为字符串列表，
            其中'Q'表示皇后的位置，'.'表示空位置。
        """
        self.result.clear()
        board = [['.' for _ in range(n)] for _ in range(n)]
        self.backtracking(0, n, board)
        return self.result
        
    def backtracking(self, row: int, n: int, board: List[List[str]]):
        """
        使用回溯算法解决N皇后问题。
        
        Args:
            row: 当前处理的行。
            n: 棋盘的大小。
            board: 当前棋盘状态。
            
        Returns:
            None: 结果将添加到self.result中。
        """
        # 终止条件
        if row == n:
            self.result.append(["".join(row) for row in board])
            return 
        # 遍历
        for col in range(n):
            diagonal = row - col
            backdiagonal = row + col
            if col in self.cols or diagonal in self.diagonals or backdiagonal in self.back_diagonals:
                continue
            board[row][col] = 'Q'
            self.cols.add(col)
            self.diagonals.add(diagonal)
            self.back_diagonals.add(backdiagonal)
            self.backtracking(row+1, n, board)
            # 回溯
            board[row][col] = '.'
            self.cols.remove(col)
            self.diagonals.remove(diagonal)
            self.back_diagonals.remove(backdiagonal)

if __name__ == "__main__":
    solution = Solution()
    result = solution.solveNQueens(4)
    print(result)

File: cpp_exercise/backtracking_2_2.cpp
/*
51.N皇后
*/
#include <iostream>
#include <set>
#include <string>
#include <vector>
using namespace std;

/**
 * @brief A solution to the N-Queens problem
 *
 * This class solves the N-Queens problem using backtracking
 */
class Solution {
public:
  /**
   * @brief 返回所有N皇后的情况
   *
   * @param n 皇后的数量
   * @return A vector of vector of strings
   */
  vector<vector<string>> solveNQueens(int n) {
    vector<string> board(n, string(n, '.'));
    backtracking(0, n, board);
    return result;
  }

private:
  set<int> backdiagonals;
  set<int> diagonals;
  set<int> cols;
  vector<vector<string>> result;
  /**
   * @ brief 回溯算法
   *
   * @param row 深搜到了第几层
   * @param n 总层
   * @param board 棋盘
   */

  void backtracking(int row, int n, vector<string> &board) {
    if (row == n) {
      result.push_back(board);
      return;
    }

    for (int col = 0; col < n; ++col) {
      int diagonal = row - col;
      int backdiagonal = row + col;
      if (cols.find(col) != cols.end() ||
          diagonals.find(diagonal) != diagonals.end() ||
          backdiagonals.find(backdiagonal) != backdiagonals.end()) {
        continue;
      } else {
        board[row][col] = 'Q';
        cols.insert(col);
        diagonals.insert(diagonal);
        backdiagonals.insert(backdiagonal);
        backtracking(row + 1, n, board);
        board[row][col] = '.';
        cols.erase(col);
        diagonals.erase(diagonal);
        backdiagonals.erase(backdiagonal);
      }
    }
  }
};

int main() {
  Solution obj;
  vector<vector<string>> result = obj.solveNQueens(4);
  for (const auto &tmp : result) {
    for (const auto &tmp2 : tmp) {
      cout << tmp2 << endl;
    }
    cout << endl;
  }
}
